nucliadb_node/nucliadb_node/indexer.py

```diff
@@ -0,0 +1,337 @@
+        instance.index_message = index_messsage
+        instance.nats_msg = nats_msg
+        instance.mpu = mpu
+        return instance
+
+    @classmethod
+    def from_msg(cls, msg: Msg) -> "WorkUnit":
+        mpu = MessageProgressUpdater(msg, nats_consumer_settings.nats_ack_wait * 0.66)
+
+        pb = IndexMessage()
+        pb.ParseFromString(msg.data)
+
+        return cls.__new__(index_messsage=pb, nats_msg=msg, mpu=mpu)
+
+    @property
+    def seqid(self) -> int:
+        return int(self.nats_msg.reply.split(".")[5])
+
+    @property
+    def node_id(self) -> str:
+        return self.index_message.node
+
+    @property
+    def shard_id(self) -> str:
+        return self.index_message.shard
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, WorkUnit):
+            return NotImplemented
+        return self._priority_id.__eq__(other._priority_id)
+
+    @cached_property
+    def _priority_id(self) -> tuple[int, int]:
+        # Priority based on message source and smaller seqid
+        source_priority = self.priorities[self.index_message.source]
+        return (source_priority, self.seqid)
+
+    def __lt__(self, other):
+        if not isinstance(other, WorkUnit):
+            return NotImplemented
+        return self._priority_id.__lt__(other._priority_id)
+
+
+class ConcurrentShardIndexer:
+    def __init__(self, writer: Writer):
+        self.writer = writer
+        self.indexers: dict[str, tuple[PriorityIndexer, asyncio.Task]] = {}
+
+    async def initialize(self):
+        self.storage = await get_storage(service_name=SERVICE_NAME)
+
+    async def finalize(self):
+        for _, task in self.indexers.items():
+            try:
+                await task
+            except Exception as exc:
+                logger.exception(
+                    "Indexer task raised an error while finalizing it", exc_info=exc
+                )
+        self.indexers.clear()
+
+    async def index_message_soon(self, msg: Msg):
+        work = WorkUnit.from_msg(msg)
+        await work.mpu.start()
+
+        logger.info(
+            f"Index message for shard {work.shard_id} is being enqueued",
+            extra={
+                "shard": work.shard_id,
+                "subject": msg.subject,
+                "reply": msg.reply,
+                "seqid": work.seqid,
+                "storage_key": work.index_message.storage_key,
+            },
+        )
+
+        await self.push_work_to_indexer(work)
+        CONCURRENT_INDEXERS_COUNT.set(
+            len(self.indexers), labels=dict(node=work.node_id)
+        )
+
+    async def push_work_to_indexer(self, work: WorkUnit):
+        exists = work.shard_id in self.indexers
+        if exists:
+            indexer, _ = self.indexers[work.shard_id]
+            await indexer.index_soon(work)
+        else:
+            indexer = PriorityIndexer(writer=self.writer, storage=self.storage)
+            await indexer.index_soon(work)
```

Comment:
Set `self.indexers[work.shard_id] = (indexer, task)` immediately after you create it here so that the next msg that comes in uses the cached in instead of potentially creating multiple at the same time before setting it on the dict

Reply:
`index_soon` is now sync, so there's no risk on adding first the work